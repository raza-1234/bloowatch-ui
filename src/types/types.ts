export enum ModalName {
  REGISTER_USER = "Register",
  LOGIN_USER = "Login",
  EDIT_USER = "Edit User"
}

export const STATUS_TEXT = "OK"
export const EXPIRED_TOKEN = "token expire" 
export const PRODUCT_NOT_FOUND = "No Product Exist."
export const INCORRECT_PASSWORD = "Wrong Password."
export const INVALID_COUPON = "Invalid coupon."
export const COUPON_APPLIED = "Coupon Applied"


// type for form 
export type RegisterLoginForm = {
  name?: string,
  email: string,
  password: string,
  confirmPassword?: string
}

//type for product 
export type Product = {
  cartProducts: CartList[]
  id: number,
  image: string,
  price: number, 
  title: string,
  quantity: number,
  category: string[]
}

//type for paging data object
export type Paging = {
  currentDataCount: number,
  currentPage: number,
  limit: number,
  moreData: boolean,
  nextPage: number,
  totalCount: number,
  totalPage: number
}

//type for cartList
export type CartList = {
  id: number,
  productId: number,
  quantity: number,
  userId: number,
  createdAt: string,
  updatedAt: string,
  product: Product
}

//user data 
export type EditUserFormValue = {
  name: string,
  email: string,
  currentPassword: string,
  newPassword: string,
  confirmPassword: string
}

//coupon detail
export type CouponDetail = {
  name: string,
  discountPercentage: number
}

//CouponFormValue
export type CouponFormValue = {
  coupon: string
}

//types for children component in context
export type Children = {
  children: JSX.Element | JSX.Element[]
}

//types for cart context
export type CartContextType = {
  cartData: CartList[]
  cartCount?: number
}

//types for cart context provider
export type CartContextProvider = {
  cart: CartContextType
  fetchCartProducts: (accessToken?: string | null, userId?: number) => Promise<void>
}

// type for decode token
export type Decoded_Token = {
  userId: number
}

//type for user detail
export type UserDetail = {
  name?: string,
  email?: string,
  id?: number
  accessToken: string | null
}

//type for auth context
export type AuthContexType = {
  getUserData: (authToken: string, userId: number) => Promise<UserDetail|void>, 
  userData: UserDetail
  setUserData: (value: UserDetail) => void
}

//types for parameter of fetchproduct function
export type QueryParam = {
  page?: string
  price?: string[]
  category?: string
  search?: string
}

//return type for fetchproduct function
export type DataDetail = {
  data: Product[],
  paging: Paging
}