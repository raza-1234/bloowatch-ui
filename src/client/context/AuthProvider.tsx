import { createContext, useState, useEffect, useContext } from "react";
import api from "../axios/api";
import { AxiosResponse } from "axios";
import React from "react";
import { jwtDecode } from "jwt-decode";
import { Decoded_Token, UserDetail, Children, AuthContexType } from "../types/types";

const AuthContext = createContext<AuthContexType | undefined>(undefined);

export const AuthProvider = ({children}: Children) => {

  const [userData, setUserData] = useState<UserDetail>({
    accessToken: localStorage.getItem("access_token") || null,
    email: "",
    id: undefined,
    name: ""
  })

  useEffect(() => {      
    if (userData?.accessToken){
      const data: any = jwtDecode(userData.accessToken);
      if ( data.userId ) {        
        getUserData(userData.accessToken, data.userId);
      }
    }
  }, [userData?.accessToken])

  const getUserData = async (accessToken: string | null, userId?: number): Promise<UserDetail|void> => {
    try {
      const response: AxiosResponse = await api.get(`user-detail/${userId}`,
        { 
          headers: {
            "Authorization" : `Bearer ${accessToken}`
          }
        }
      )      
      setUserData({...userData, ...response.data});
    } catch (err){
      console.log(err);
    }
  }

  return(
    <AuthContext.Provider value={{ getUserData, userData, setUserData}}>
      {children}
    </AuthContext.Provider>
  )
}

const useAuthData = () => {
  const context = useContext(AuthContext)
  if(!context) {
    throw new Error('AuthData must in the AuthProvider')
  }

  return context;
}

export default useAuthData