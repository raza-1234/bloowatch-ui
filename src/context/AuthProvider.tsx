import { createContext, useState, useEffect, useContext } from "react";
import api from "../axios/api";
import { AxiosResponse } from "axios";
import React from "react";
import { jwtDecode } from "jwt-decode";
import { Decoded_Token, UserDetail, Children, AuthContexType } from "../types/types";

const AuthContext = createContext<AuthContexType | undefined>(undefined);

export const AuthProvider = ({children}: Children) => {

  const [userData, setUserData] = useState<UserDetail>({
    accessToken: localStorage.getItem("access_token") || null,
    email: "",
    id: undefined,
    name: ""
  })

  useEffect(() => {  
    if (userData?.accessToken){
      const { userId }: Decoded_Token = jwtDecode(userData.accessToken);
      if ( userId ) {
        getUserData(userData.accessToken, userId);
      }
    }
  }, [userData?.accessToken])

  const getUserData = async (accessToken: string, userId: number): Promise<UserDetail|void> => {
    try {
      const response: AxiosResponse = await api.get(`http://localhost:3500/user-detail/${userId}`,
        { 
          headers: {
            "Authorization" : `Bearer ${accessToken}`
          }
        }
      )      
      setUserData({...userData, ...response.data});
    } catch (err){
      console.log(err);
    }
  }

  return(
    <AuthContext.Provider value={{ getUserData, userData, setUserData}}>
      {children}
    </AuthContext.Provider>
  )
}

const AuthData = () => {
  const context = useContext(AuthContext)
  console.log('context', context)
  if(!context) {
    throw new Error('AuthData must in the AuthProvider')
  }

  return context;
}

export default AuthData