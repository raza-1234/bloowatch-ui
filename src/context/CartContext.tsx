import React, {createContext, useContext, useEffect, useState} from "react";
import { AxiosResponse } from 'axios';
import { Children, CartContextType, CartContextProvider, STATUS_TEXT } from "../types/types";
import api from '../axios/api';
import { jwtDecode } from "jwt-decode";
import AuthData from "./AuthProvider";

const CartContext = createContext<CartContextProvider | null>(null);

export const CartProvider = ({children}: Children) => {
  const [cart, setCart] = useState<CartContextType>({
    cartData: [],
    cartCount: 0,
  })
  
  const { userData } = AuthData();

  useEffect(() => {   
    if (userData?.accessToken){
      const { userId }: {userId: number} = jwtDecode(userData?.accessToken);
      if ( userId ) {
        fetchCartProducts(userData?.accessToken, userId)
      }
    } 
  },[userData?.accessToken])

  const fetchCartProducts = async (accessToken?: string, userId?: number): Promise<void> => {
    try {
      const response: AxiosResponse = await api.get(`cart/getAllCartProducts/${userId}`, 
        {
          headers: {"Authorization" : `Bearer ${accessToken}`
        }
      })      
      if (response.statusText === STATUS_TEXT && response.data.length){
        setCart({cartData: response.data, cartCount: response.data.length});
        return;
      }
      setCart({cartData: [], cartCount: 0});
    } catch (err: any){
      console.log(err);
    }
  }

  return (
    <CartContext.Provider value = {{cart, fetchCartProducts}} >
      {children}
    </CartContext.Provider>
  )
}

const CartContextData = () => {
  const context = useContext(CartContext);
  if(!context) {
    throw new Error('Cart Context Not Setup Properly.')
  }

  return context;
}
export default CartContextData;


