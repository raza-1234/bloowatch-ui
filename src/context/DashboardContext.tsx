import React, { FC, createContext, useState} from 'react'
import { Product, Paging, PRODUCT_STATUS, RESPONSE_TEXT, DashboardContextValue, Children } from '../types/types';
import { AxiosResponse } from 'axios';
import api from '../axios/api';
import useAuth from '../hooks/useAuth';
import logOut from '../utils/logout';

export const dashboardContext = createContext<DashboardContextValue | null>(null);
const DashboardContext = ({children}: Children) => {

  const [ products, setProducts ] = useState<Product[]>([])
  const [pagingInfo, setPagingInfo] = useState<Paging>()
  const [page, setPage] = useState<number>();
  const [category, setCategory] = useState("")
  const [search, setSearch] = useState("")
  const [price, setPrice] = useState<number[]>([0, 100]);
  const { auth, setAuth }: any = useAuth();  

  async function fetchProducts(productPrice: number[], productCategory?: string, title?: string, pageNumber?: number): Promise<void> {       
    try {
      let response: AxiosResponse;
      if (title?.trim()){
        response = await api.get(`products/get-products?price[]=${productPrice[0]}&price[]=${productPrice[1]}&page=${pageNumber}&search=${title.toLowerCase()}`, {headers: {"Authorization" : `Bearer ${auth.token}`}}
        );
      }
      else if (productCategory?.trim()){
        response = await api.get(`products/get-products?price[]=${productPrice[0]}&price[]=${productPrice[1]}&page=${pageNumber}&category=${productCategory.toLowerCase()}`, {headers: {"Authorization" : `Bearer ${auth.token}`}});
      }
      else {
        response = await api.get(`products/get-products?price[]=${productPrice[0]}&price[]=${productPrice[1]}&page=${pageNumber}`, {headers: {"Authorization" : `Bearer ${auth.token}`}});
      }

      if (response.data.message === PRODUCT_STATUS){
        setProducts([])
        setPagingInfo(undefined)
        return ; 
      }            
      setProducts(response.data.data)
      setPagingInfo(response.data.paging)
    } catch (err){
      console.log(err);
      if (err.response.data === RESPONSE_TEXT){
        logOut();
        setAuth();
      }
    }
  }  

  return (
    <dashboardContext.Provider 
      value={
        {products, pagingInfo, fetchProducts, page, setPage, category, setCategory, search, setSearch, setPrice, price  }
      }
    >
      {children}
    </dashboardContext.Provider>
  )
}

export default DashboardContext
