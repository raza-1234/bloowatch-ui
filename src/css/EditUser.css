.bloowatch-edit-user__wrapper {
  margin: 60px 0 10px 0;
  display: flex;
  align-items: center;
  justify-content: center;
}

.bloowatch-edit-user__form {
  position: relative;
  text-align: left;
  width: 400px ;
  padding: 20px 20px 30px 20px;
  margin: 5px 0px;
  border-radius: 3px;
  box-shadow: 1px 1px 10px 5px rgba(0, 0, 0,0.2);
}

/* style for name field */
.bloowatch-edit-user__user-name {
  display: flex;
  align-items: center;
  position: relative;
  margin-top: 10px;
}
.bloowatch-edit-user__user-name input{
  width: 100%;
  border: 1px solid rgb(235, 235, 235);
  padding: 8px 8px 8px 25px;
  border-radius: 4px ;
}
.bloowatch-edit-user__name-icon {
  position: absolute;
  left: 5px;
  color: rgb(211, 211, 211);
  font-size: 13px;
}

/* style for password field */

.bloowatch-edit-user__user-password {
  display: flex;
  align-items: center;
  border: 1px solid rgb(235, 235, 235);
  border-radius: 4px ;
  width: 100%;
  margin-top: 10px; 
  color: var(--ternary);
  position: relative;
}

.bloowatch-edit-user__user-password input {
  width: 100%;
  border: 0px; 
  margin-top: 0px;
  border-radius: 4px ;
  padding: 8px 8px 8px 25px;
}

.bloowatch-edit-user__user-password p{
  position: absolute;
  left: 88%;
  font-size: 13px;
  cursor: pointer;
}

.bloowatch-edit-user__password-icon {
  position: absolute;
  left: 5px;
  color: rgb(211, 211, 211);
}

.bloowatch-edit-user__button {
  width: 100%;
  margin-top: 10px;
}

.bloowatch-edit-user__button button{
  display: flex;
  align-items: center;
  justify-content: center;
  gap: 10px;
  color: var(--secondary);
  font-weight: 800;
  width: 100%;
  padding: 10px;
  background-color: var(--primary);
  border: 0px;
  border-radius: 4px;
  cursor: pointer;
}


