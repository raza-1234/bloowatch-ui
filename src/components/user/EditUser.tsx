import "../../css/EditUser.css"
import React, { useState } from 'react';
import { useForm } from 'react-hook-form';
import { IoMdLock } from "react-icons/io";
import { FaUser } from "react-icons/fa6";
import { MdEmail } from 'react-icons/md';

import { ModalName, EditUserFormValue, INCORRECT_PASSWORD, STATUS_TEXT } from '../../types/types';
import { validationError } from '../../utils/validationError';
import { AxiosResponse } from 'axios';
import api from '../../axios/api';
import { successAlert, errorAlert } from '../../utils/toast';
import AuthData from '../../context/AuthProvider';

const EditUser = () => {

  const { getUserData, userData } = AuthData();

  const [showOldPassword, setShowOldPassword] = useState(false);
  const [showNewPassword, setShowNewPassword] = useState(false);
  const [showConfirmPassword, setShowConfirmPassword] = useState(false); 
  const [changePassword, setChangePassword] = useState(false);

  const form = useForm<EditUserFormValue>()
  const { register, handleSubmit, formState, resetField, watch } = form;
  const {errors} = formState; 

  const submitHandler = async (data: EditUserFormValue): Promise<void> =>{    
    try {
      const response: AxiosResponse = await api.put("http://localhost:3500/edit-user", 
      {
        ...data, confirmPassword: undefined 
      }, 
      {
        headers: {
          "Authorization" : `Bearer ${userData?.accessToken}`
        }
      })

      if (response.statusText === STATUS_TEXT){
        successAlert(response.data.message)

        await getUserData(userData?.accessToken!, userData?.id!)

        resetField("currentPassword")
        resetField("newPassword")
        resetField("confirmPassword")
      }
    } catch (err){
      console.log(err);
      if (err.response.data.message === INCORRECT_PASSWORD){        
        return errorAlert(INCORRECT_PASSWORD)
      }
    }
  }

  return (
    <div data-testid = "form__wrapper" className='bloowatch-edit-user__wrapper'>
    <form data-testid = "form" className='bloowatch-edit-user__form' onSubmit={handleSubmit(submitHandler)} noValidate>
    <h3>{ModalName.EDIT_USER}</h3>

    <div className='bloowatch-edit-user__user-name'>
    <FaUser className='bloowatch-edit-user__name-icon'/>
      <input
        type='text'
        placeholder='Name'
        defaultValue={userData?.name}
        {
          ...register("name", 
            {
              required: "name is required.",
              validate: (val: string) => {
                if (!val.trim()) {
                  return "name can not be empty.";
                }
              }
            }
          )
        }
      />
    </div>
    {validationError(errors.name?.message)}

    <div className='bloowatch-register-login__user-email'>
      <MdEmail className='bloowatch-register-login__email-icon'/>
      <input
        readOnly
        type='email'
        placeholder='Email'
        defaultValue={userData?.email}
        {
          ...register("email",
          )
        }       
      />
      </div>

      <div className='bloowatch-edit-user__user-password'>
        <IoMdLock className='bloowatch-edit-user__password-icon'/>
        <input 
          type= { showOldPassword ? "text" : "password"}
          placeholder='Current Password'
          defaultValue=""
          {
            ...register("currentPassword",
              {
                required: "current password is required.",
                validate: (val: string) => {
                  if (!val.trim()) {
                    return "current password can not be empty.";
                  }
                }
              }
            )
          }
        />
        <p onClick={() => setShowOldPassword(!showOldPassword)}>{showOldPassword? "Hide": "Show"}</p>          
      </div>
      {validationError(errors.currentPassword?.message)}

      {
        changePassword && 
        <React.Fragment>
          <div className='bloowatch-edit-user__user-password'>
          <IoMdLock className='bloowatch-edit-user__password-icon'/>
          <input 
            type= { showNewPassword ? "text" : "password"}
            placeholder='New Password'
            defaultValue=""
            {
              ...register("newPassword",
              {
                required: "new password is required.",
                validate: (val: string) => {
                  if (!val.trim()) {
                    return "new password can not be empty.";
                  }
                }
              }
              )
            }
          />
          <p onClick={() => setShowNewPassword(!showNewPassword)}>{showNewPassword? "Hide": "Show"}</p>          
          </div>
          {validationError(errors.newPassword?.message)}

          <div className='bloowatch-edit-user__user-password'>
          <IoMdLock className='bloowatch-edit-user__password-icon'/>
          <input 
            type= { showConfirmPassword ? "text" : "password"}
            placeholder='Confirm Password'
            defaultValue=""
            {
              ...register("confirmPassword",
                {
                  required: "confirm password is required.",
                  validate: (val: string) => {
                    if (watch('newPassword') !== val) {
                      return "Your new-password and confirm-password do no match";
                    }
                  }
                }
              )
            }
          />
          <p onClick={() => setShowConfirmPassword(!showConfirmPassword)}>{showConfirmPassword? "Hide": "Show"}</p>          
          </div>
          {validationError(errors.confirmPassword?.message)}
        </React.Fragment>
      }
      
      <div className='bloowatch-edit-user__button'>
        <button>Save Changes</button>
        {
          changePassword ? 
          <button type='button' onClick={() => setChangePassword(!changePassword)}>Hide Password</button>
          :<button type='button' onClick={() => setChangePassword(!changePassword)}>Change Password</button>
        }
      </div>
    </form>
    </div>
      
  )
}

export default EditUser

