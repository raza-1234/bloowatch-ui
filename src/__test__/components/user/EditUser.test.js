import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import EditUser from "../../../components/user/EditUser"
import { BrowserRouter } from 'react-router-dom';
import { AuthProvider } from '../../../context/AuthProvider';
import { CartProvider } from "../../../context/CartContext"
import { ModalName } from '../../../types/types';

const buildApp = () => {
  return (
    render(
      <BrowserRouter>
        <AuthProvider>
          <CartProvider>
            <EditUser/>
          </CartProvider>
        </AuthProvider>
      </BrowserRouter>
    )
  )
}

describe("EditUser", () => {

  it("should render successfully", () => {
    buildApp();

    const form__wrapper = screen.getByTestId("form__wrapper");
    const form = screen.getByTestId("form");

    expect(form__wrapper).toBeInTheDocument();
    expect(form__wrapper).toHaveClass("bloowatch-edit-user__wrapper"); //selecting with class giving an erorr in development...
    expect(form).toBeInTheDocument();
    expect(form).toHaveClass("bloowatch-edit-user__form");
    expect(screen.getByRole("heading", {name: ModalName.EDIT_USER})).toBeInTheDocument();
  })

  it("should render form input fields", () => {
    const { container } = buildApp();

    expect(container).toMatchSnapshot();
  })

})