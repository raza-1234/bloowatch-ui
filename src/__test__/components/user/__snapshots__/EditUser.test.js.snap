// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditUser should render form input fields 1`] = `
<div>
  <div
    class="bloowatch-edit-user__wrapper"
    data-testid="form__wrapper"
  >
    <form
      class="bloowatch-edit-user__form"
      data-testid="form"
      novalidate=""
    >
      <h3>
        Edit User
      </h3>
      <div
        class="bloowatch-edit-user__user-name"
      >
        <svg
          class="bloowatch-edit-user__name-icon"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M224 256A128 128 0 1 0 224 0a128 128 0 1 0 0 256zm-45.7 48C79.8 304 0 383.8 0 482.3C0 498.7 13.3 512 29.7 512H418.3c16.4 0 29.7-13.3 29.7-29.7C448 383.8 368.2 304 269.7 304H178.3z"
          />
        </svg>
        <input
          name="name"
          placeholder="Name"
          type="text"
          value=""
        />
      </div>
      <p
        class="bloowatch-login-register__error"
      />
      <div
        class="bloowatch-register-login__user-email"
      >
        <svg
          class="bloowatch-register-login__email-icon"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M20 4H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2zm0 4l-8 5-8-5V6l8 5 8-5v2z"
          />
        </svg>
        <input
          name="email"
          placeholder="Email"
          readonly=""
          type="email"
          value=""
        />
      </div>
      <div
        class="bloowatch-edit-user__user-password"
      >
        <svg
          class="bloowatch-edit-user__password-icon"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M376 186h-20v-40c0-55-45-100-100-100S156 91 156 146v40h-20c-22.002 0-40 17.998-40 40v200c0 22.002 17.998 40 40 40h240c22.002 0 40-17.998 40-40V226c0-22.002-17.998-40-40-40zM256 368c-22.002 0-40-17.998-40-40s17.998-40 40-40 40 17.998 40 40-17.998 40-40 40zm62.002-182H193.998v-40c0-34.004 28.003-62.002 62.002-62.002 34.004 0 62.002 27.998 62.002 62.002v40z"
          />
        </svg>
        <input
          name="currentPassword"
          placeholder="Current Password"
          type="password"
          value=""
        />
        <p>
          Show
        </p>
      </div>
      <p
        class="bloowatch-login-register__error"
      />
      <div
        class="bloowatch-edit-user__button"
      >
        <button>
          Save Changes
        </button>
        <button
          type="button"
        >
          Change Password
        </button>
      </div>
    </form>
  </div>
</div>
`;
